x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java

```diff
@@ -1890,16 +1907,62 @@ public BiConsumer<DiscoveryNode, ClusterState> getJoinValidator() {
     @Override
     public void reload(Settings settings) throws Exception {
         if (enabled) {
-            realms.get().stream().filter(r -> JwtRealmSettings.TYPE.equals(r.realmRef().getType())).forEach(realm -> {
-                if (realm instanceof JwtRealm jwtRealm) {
-                    jwtRealm.rotateClientSecret(
-                        CLIENT_AUTHENTICATION_SHARED_SECRET.getConcreteSettingForNamespace(realm.realmRef().getName()).get(settings)
-                    );
-                }
-            });
+            final List<Exception> reloadExceptions = new ArrayList<>();
+            try {
+                reloadRemoteClusterCredentials(settings);
+            } catch (Exception ex) {
+                reloadExceptions.add(ex);
+            }
+
+            try {
+                reloadSharedSecretsForJwtRealms(settings);
+            } catch (Exception ex) {
+                reloadExceptions.add(ex);
+            }
+
+            if (false == reloadExceptions.isEmpty()) {
+                final var combinedException = new ElasticsearchException(
+                    "secure settings reload failed for one or more security components"
+                );
+                reloadExceptions.forEach(combinedException::addSuppressed);
+                throw combinedException;
+            }
+        } else {
+            ensureNoRemoteClusterCredentialsOnDisabledSecurity(settings);
         }
     }
 
+    private void reloadSharedSecretsForJwtRealms(Settings settingsWithKeystore) {
+        getRealms().stream().filter(r -> JwtRealmSettings.TYPE.equals(r.realmRef().getType())).forEach(realm -> {
+            if (realm instanceof JwtRealm jwtRealm) {
+                jwtRealm.rotateClientSecret(
+                    CLIENT_AUTHENTICATION_SHARED_SECRET.getConcreteSettingForNamespace(realm.realmRef().getName()).get(settingsWithKeystore)
+                );
+            }
+        });
+    }
+
+    /**
+     * This method uses a transport action internally to access classes that are injectable but not part of the plugin contract.
+     * See {@link TransportReloadRemoteClusterCredentialsAction} for more context.
+     */
+    private void reloadRemoteClusterCredentials(Settings settingsWithKeystore) {
+        final PlainActionFuture<ActionResponse.Empty> future = new PlainActionFuture<>() {
+            @Override
+            protected boolean blockingAllowed() {
+                // Accepting a blocking call here since the underlying action is local-only and only performs fast in-memory ops
+                // (extracts a subset of passed in `settingsWithKeystore` and stores them in a map)
+                return true;
+            }
+        };
+        getClient().execute(
+            ReloadRemoteClusterCredentialsAction.INSTANCE,
+            new ReloadRemoteClusterCredentialsAction.Request(settingsWithKeystore),
+            future
+        );
+        future.actionGet();
```

Comment:
AIUI we're not blocking here, the future will always be complete at this point, so we could instead do this:

```suggestion
        assert future.isDone();
        future.actionGet(0, TimeUnit.NANOSECONDS);
```

A zero-timeout get is permitted on this thread without having to override `blockingAllowed()`.

Reply:
Nice, TIL! Thanks for the pointer, I'll validate but this should indeed work for this case.

Reply:
Got it, thanks for confirming!